### ISSUE TITLE
`@abstractmethod` disables specialized error message

### ISSUE THREAD

**Issue by DanielRosenwasser** [microsoft/pyright#5936]
## Environment data

-   Language Server version: 2023.9.10 (pyright c70adefc)
-   OS and version: Windows 11
-   Python version (& distribution if applicable, e.g. Anaconda): 3.10.11

## Code Snippet

```python
from abc import ABC, abstractmethod

class Thing(ABC):
    def method_a(input: str) -> str:
    #            ~~~~~~~~~~
    # error: Instance methods should take a "self" parameter ✅
        ...

    @abstractmethod
    def method_b(input: str) -> str:
    #                   ~~~
    # error: Type of parameter "input" must be a supertype of its class "Thing" ❌
        ...
```

## Expected behavior

Both methods have good errors suggesting to define a `self` parameter.

## Actual behavior

Only the non-decorated method gets the specialized error.

**Comment by erictraut** [comment]
Could someone from the pylance team please transfer this to the pyright project since it's part of the core type checker? Thanks!

**Comment by erictraut** [comment]
@DanielRosenwasser, pyright currently skips the check for `self` or `cls` any time an override is applied to a method. Unlikely TypeScript, decorators in Python are allowed to completely mutate the signature of the decorated method — or replace them with any other object, for that matter. For that reason, pyright needs to back off some of its non-type-related checks when decorators are present. I suppose that we could special-case a few known stdlib decorators like `@abstractmethod`, but I'm not sure what principle we'd use for deciding which to special case and which to not. It might be best to leave the current logic in place.

**Comment by DanielRosenwasser** [comment]
I figured it was something along those lines. I think a common case for built-ins would definitely be helpful for what it's worth.

**Comment by erictraut** [comment]
I think that special-casing builtins is reasonable, but there are only two decorators in the builtins module: `@staticmethod` and `@classmethod`. Those are already handled as special cases. 

`@abstractmethod` isn't a builtin. It's exported from the `abc` module. Special-casing all decorators exported from the stdlib doesn't strike me as practical since there are hundreds of them.

One could argue that pyright (and pylance) shouldn't enforce the name of the `self` or `cls` parameter in the first place, since this isn't a type checking concern. It's more of a linter concern. Maybe it would be best to remove it altogether? I personally find it useful to have in pyright/pylance, but that might not be a good enough reason to keep it.

**Comment by erictraut** [comment]
I think we'll leave the behavior as is for now.
